### PR TITLE
feat: update cosmwasm to 2.0 #NTRN-302

### PR DESCRIPTION
**Pull request by joldie777**
Task: https://hadronlabs.atlassian.net/browse/NTRN-302

Related PRs:
https://github.com/neutron-org/neutron-dev-contracts/pull/54

List of changes:
1. Due to [this update](https://github.com/CosmWasm/cosmwasm/blob/main/MIGRATING.md#:~:text=Also%20remove%20any%20uses%20of%20the%20backtraces%20feature), any uses of the `backtraces` feature have been removed.
2. Due to [this update](https://github.com/CosmWasm/cosmwasm/blob/main/MIGRATING.md#:~:text=you%20now%20have%20to%20explicitly%20call%20new), `new` is now explicitly called to create `Binary` instance from an inner value.
3. Due to [this update](https://github.com/CosmWasm/cosmwasm/blob/main/MIGRATING.md#:~:text=functions%20were%20removed%20from%20the%20MockQuerier), newly exposed module is used to access the `update_staking` function directly.
4. Due to [this update](https://github.com/CosmWasm/cosmwasm/blob/main/MIGRATING.md#:~:text=In%20test%20code%2C%20replace%20calls%20to%20mock_info%20with%20message_info), calls to `mock_info` has been replaced with `message_info`.
5. Due to [non_exhaustive attribute](https://github.com/CosmWasm/cosmwasm/blob/v2.0.4/packages/std/src/query/staking.rs#L65), the `Delegation` struct is no longer accessible from the outside of the crate where it defined. So [StdDelegation](https://github.com/neutron-org/neutron-sdk/blob/586adb91fe1b30ecaf510f975ce13c2bc39d7431/packages/neutron-sdk/src/interchain_queries/v045/types.rs#L447) struct was created as the copy of the cosmwasm `Delegation` to be used instead.
6. Due to [deprecation](https://github.com/CosmWasm/cosmwasm/blob/v2.0.4/packages/std/src/results/submessages.rs#L274), the uses of the `data` field have been replaced with `msg_responses`. Since `msg_responses` contains protobuf-encoded messages, the JSON deserialization methods have been replaced by `decode_message_response` with corresponding replacement of the response types: [MsgIbcTransferResponse](https://github.com/neutron-org/neutron-sdk/blob/main/packages/neutron-sdk/src/bindings/msg.rs#L568) for [MsgTransferResponse](https://github.com/neutron-org/neutron-sdk/blob/main/packages/neutron-sdk/src/proto_types/neutron.transfer.rs#L35) and [MsgSubmitTxResponse](https://github.com/neutron-org/neutron-sdk/blob/main/packages/neutron-sdk/src/bindings/msg.rs#L558) for [MsgSubmitTxResponse](https://github.com/neutron-org/neutron-sdk/blob/main/packages/neutron-sdk/src/proto_types/neutron.interchaintxs.v1.rs#L96)

### CODE REVIEW

**Comment by joldie777** [comment]
Test run: https://github.com/neutron-org/neutron-tests/actions/runs/9489743999

**Comment by joldie777** [comment]
Test run: https://github.com/neutron-org/neutron-tests/actions/runs/9515290655

File: contracts/ibc_transfer/src/contract.rs [hunk 1]
```diff
@@ -112,14 +114,18 @@ fn msg_with_sudo_callback<C: Into<CosmosMsg<T>>, T>(
...
     let payload = read_reply_payload(deps.storage, msg.id)?;
-    let resp: MsgIbcTransferResponse = from_json(
-        msg.result
+    let resp: MsgTransferResponse = decode_message_response(
+        &msg.result
             .into_result()
             .map_err(StdError::generic_err)?
-            .data
-            .ok_or_else(|| StdError::generic_err("no result"))?,
+            .msg_responses[0]
```
Comment (pr0n00gler) [inline contracts/ibc_transfer/src/contract.rs:121]: Could you please add a comment why do we use only the first element here?

File: contracts/ibc_transfer/src/contract.rs [hunk 2]
```diff
@@ -112,14 +114,18 @@ fn msg_with_sudo_callback<C: Into<CosmosMsg<T>>, T>(
...
+            .clone()
+            .value
+            .to_vec(),
     )
     .map_err(|e| StdError::generic_err(format!("failed to parse response: {:?}", e)))?;
+    deps.api
```
Comment (pr0n00gler) [inline contracts/ibc_transfer/src/contract.rs:127]: I think this is redundant

File: contracts/neutron_interchain_txs/src/contract.rs [hunk 3]
```diff
@@ -615,13 +616,14 @@ fn sudo_error(deps: DepsMut, request: RequestPacket, details: String) -> StdResu
...
-    let resp: MsgSubmitTxResponse = serde_json_wasm::from_slice(
-        msg.result
+    let resp: MsgSubmitTxResponse = decode_message_response(
+        &msg.result
             .into_result()
             .map_err(StdError::generic_err)?
-            .data
-            .ok_or_else(|| StdError::generic_err("no result"))?
-            .as_slice(),
+            .msg_responses[0]
```
Comment (pr0n00gler) [inline contracts/neutron_interchain_txs/src/contract.rs:623]: The same about `.msg_responses[0]`

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: contracts/ibc_transfer/src/contract.rs (hunk 1 above)
Comment (joldie777) [reply contracts/ibc_transfer/src/contract.rs]: Done

File: contracts/ibc_transfer/src/contract.rs (hunk 2 above)
Comment (joldie777) [reply contracts/ibc_transfer/src/contract.rs]: Removed

File: contracts/neutron_interchain_txs/src/contract.rs (hunk 3 above)
Comment (joldie777) [reply contracts/neutron_interchain_txs/src/contract.rs]: Done

**Comment by quasisamurai** [comment]
please merge `main`, i see diff locally

**Review by pr0n00gler — DISMISSED** [review]

**Review by sotnikov-s — APPROVED** [review]
please remove several `backtraces` occurrences in the project. the rest LGTM

**Review by sotnikov-s — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Pull request merged**